src/lib/pubkey/dilithium/dilithium.cpp

```diff
@@ -0,0 +1,749 @@
+      std::vector<uint8_t> m_msg;
+
+   };
+
+   class Dilithium_Verification_Operation final : public PK_Ops::Verification
+   {
+   public:
+      Dilithium_Verification_Operation(const Dilithium_PublicKey& pub_dilithium) : m_pub_key (pub_dilithium)
+      {
+      }
+      /*
+      * Add more data to the message currently being signed
+      * @param msg the message
+      * @param msg_len the length of msg in bytes
+      */
+      void update(const uint8_t msg[], size_t msg_len) override 
+      {
+         m_msg.insert(m_msg.end(), msg, msg + msg_len);
+      }
+
+      /*
+      * Perform a verification operation
+      * @param rng a random number generator
+      */
+      bool is_valid_signature(const uint8_t* sig, size_t sig_len) override
+      {
+          const auto& mode = m_pub_key.m_public->get_mode();
+         if (sig_len != mode.get_crypto_bytes())
+         {
+            return false;
+         }
+
+         PolynomialVector z(mode.get_l());
+         PolynomialVector h(mode.get_k());
+         std::array<uint8_t, DilithiumModeConstants::SEEDBYTES> c;
+         if (PolynomialVector::unpack_sig(c.data(), z, h, sig, mode))
+         {
+            return false;
+         }
+
+         if (z.polyvec_chknorm(mode.get_gamma1() - mode.get_beta()))
+         {
+            return false;
+         }
+
+         /* Compute CRH(CRH(rho, t1), msg) */
+         auto mu = mode.CRH(m_pub_key.m_public->get_raw_pk().data(), m_pub_key.m_public->get_raw_pk().size(),
+            mode.CRHBYTES);
+         SHAKE_256 shake256_variable(DilithiumModeConstants::CRHBYTES * 8);
+         shake256_variable.update(mu.data(), DilithiumModeConstants::CRHBYTES);
+         shake256_variable.update(m_msg.data(), m_msg.size());
+         mu = shake256_variable.final();
+
+         /* Matrix-vector multiplication; compute Az - c2^dt1 */
+         auto matrix = PolynomialMatrix::generate_matrix(m_pub_key.m_public->get_rho(), mode);
+
+         auto cp = Polynomial::poly_challenge(c.data(), mode);
+         cp.ntt();
+
+         PolynomialVector t1 = m_pub_key.m_public->get_t1();
+         t1.polyvec_shiftl();
+         t1.ntt();
+         t1.polyvec_pointwise_poly_montgomery(t1, cp);
+
+         z.ntt();
+
+         auto w1 = matrix.generate_polyvec_matrix_pointwise_montgomery(z, mode);
+         w1 -= t1;
+         w1.reduce();
+         w1.invntt_tomont();
+         w1.cadd_q();
+         w1.polyvec_use_hint(w1, h, mode);
+         auto packed_w1 = w1.polyvec_pack_w1(mode);
+
+         /* Call random oracle and verify challenge */
+         SHAKE_256 shake256_variable1(DilithiumModeConstants::SEEDBYTES * 8);
+         shake256_variable1.update(mu.data(), mode.CRHBYTES);
+         shake256_variable1.update(packed_w1.data(), packed_w1.size());
+         auto c2 = shake256_variable1.final();
+            
+         if (std::equal(c.begin(), c.end(), c2.begin()))
+         {
+            return true; 
+         }
+         return false;
```

Comment:
```suggestion
         BOTAN_ASSERT_NOMSG(c.size() == c2.size());
         return std::equal(c.begin(), c.end(), c2.begin());
```

... does this need to perform a constant-time comparison by any chance? @randombit